src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -2186,17 +2186,19 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
                 else if (boundExpr is BoundCollectionLiteralExpression convertedCollection)
                 {
                     type = null;
-                    if (highestBoundExpr is BoundConversion { ConversionKind: ConversionKind.CollectionLiteral, Conversion: var convertedCollectionConversion })
+                    if (highestBoundExpr is BoundConversion { ConversionKind: ConversionKind.CollectionLiteral or ConversionKind.NoConversion, Conversion: var convertedCollectionConversion })
                     {
-                        convertedType = convertedCollection.Type;
+                        convertedType = highestBoundExpr.Type;
                         convertedNullability = convertedCollection.TopLevelNullability;
                         conversion = convertedCollectionConversion;
                     }
                     else
                     {
-                        Debug.Assert(false); // Add test if this assert fails.
-                        (convertedType, convertedNullability) = (type, nullability);
-                        conversion = Conversion.Identity;
+                        // There was an explicit cast on top of this.
+                        (convertedType, convertedNullability) = (convertedCollection.Type, nullability);
+                        conversion = convertedCollection.CollectionTypeKind == CollectionLiteralTypeKind.None
+                            ? Conversion.NoConversion
+                            : Conversion.CollectionLiteral;
```

Comment:
It looks like the switch and conditional expressions above use an identity conversion in the corresponding position. What's different about this case that we use this conversion instead here?

Reply:
Those cases have a natural type, whereas the collection literal does not, so we need to capture whether the explicit conversion was valid (a *collection literal conversion*) or invalid (*no conversion*).